external-crates/move/crates/move-compiler/src/typing/core.rs

```diff
@@ -2162,3 +2278,119 @@ fn check_num_tvar_(subst: &Subst, ty: &Type) -> bool {
+    id1: RefVar,
+    lhs: &Type,
+    loc2: Loc,
+    id2: RefVar,
+    rhs: &Type,
+) -> Result<(Subst, Type), TypingError> {
+    use Type_::AutoRef;
+    let ref_var1 = forward_ref_var(&subst, id1);
+    let ref_var2 = forward_ref_var(&subst, id2);
+
+    let ref_kind1 = subst.get_ref_var(ref_var1);
+    let ref_kind2 = subst.get_ref_var(ref_var2);
+
+    let new_ref_var = RefVar::next();
+
+    match (ref_kind1, ref_kind2) {
+        (None, None) => {
+            // If both auto-refs are unset, combine them and join their types.
+            subst.insert_ref_var(ref_var1, RefKind::Forward(new_ref_var));
+            subst.insert_ref_var(ref_var2, RefKind::Forward(new_ref_var));
+            let (subst, ty) = join_impl(subst, case, lhs, rhs)?;
+            Ok((subst, sp(loc2, AutoRef(new_ref_var, Box::new(ty)))))
+        }
+        (Some(rk1), Some(rk2)) => {
+            assert!(!matches!(rk1, RefKind::Forward(_)));
+            assert!(!matches!(rk2, RefKind::Forward(_)));
+            if let Some(join_kind) = rk1.join(rk2) {
+                // If both auto-refs are set and can join, we join them.
+                subst.insert_ref_var(ref_var1, RefKind::Forward(new_ref_var));
+                subst.insert_ref_var(ref_var2, RefKind::Forward(new_ref_var));
+                subst.insert_ref_var(new_ref_var, join_kind);
+                let (subst, ty) = join_impl(subst, case, lhs, rhs)?;
+                Ok((subst, sp(loc2, AutoRef(new_ref_var, Box::new(ty)))))
+            } else {
+                // If they are both already set without a join, realize and recur for the error.
+                let lhs = realize_autoref(loc1, rk1, lhs.clone());
+                let rhs = realize_autoref(loc2, rk2, rhs.clone());
+                join_impl(subst, case, &lhs, &rhs)
+            }
+        }
+        (None, Some(rk2)) => {
+            let lhs = sp(loc1, AutoRef(ref_var1, Box::new(lhs.clone())));
+            let rhs = realize_autoref(loc2, rk2, rhs.clone());
+            join_impl(subst, case, &lhs, &rhs)
+        }
+        (Some(rk1), None) => {
+            let lhs = realize_autoref(loc1, rk1, lhs.clone());
+            let rhs = sp(loc2, AutoRef(ref_var2, Box::new(rhs.clone())));
+            join_impl(subst, case, &lhs, &rhs)
+        }
+    }
+}
+
+fn join_bind_ref_var(
+    mut subst: Subst,
+    case: TypingCase,
+    loc: Loc,
+    rv: RefVar,
+    rvtype: &Type,
+    other @ sp!(_, otype): &Type,
+) -> Result<(Subst, Type), TypingError> {
+    use Type_::*;
+    let rv = forward_ref_var(&subst, rv);
+
+    match subst.get_ref_var(rv) {
+        Some(RefKind::Forward(_)) => unreachable!(),
+        None => match otype {
+            Ref(false, t) => {
+                subst.insert_ref_var(rv, RefKind::ImmRef);
+                join_impl(subst, case, rvtype, t)
+            }
+            Ref(true, t) => {
+                subst.insert_ref_var(rv, RefKind::MutRef);
+                join_impl(subst, case, rvtype, t)
+            }
```

Comment:
While the case likely isn't going to happen currently (given the limited usage), I think we should harden this in the case where `otype` is a `Var(_)`. Thoughts?

Reply:
`Var(_)` and `AutoRef(_, _)` are now both marked unreachable (as they should have been handled in `join_impl`).